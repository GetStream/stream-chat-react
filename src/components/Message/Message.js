// @ts-check
import React, { useCallback, useContext } from 'react';
import PropTypes from 'prop-types';
import { ChannelContext } from '../../context';
import MessageSimple from './MessageSimple';
import { checkChannelPropType, checkClientPropType } from '../../utils';
import {
  MESSAGE_ACTIONS,
  getMessageActions,
  areMessagePropsEqual,
} from './utils';
import {
  useMuteHandler,
  useEditHandler,
  useReactionHandler,
  useDeleteHandler,
  useActionHandler,
  useRetryHandler,
  useMentionsHandler,
  useUserHandler,
  useOpenThreadHandler,
  useUserRole,
  useFlagHandler,
} from './hooks';

/**
 * Message - A high level component which implements all the logic required for a message.
 * The actual rendering of the message is delegated via the "Message" property
 *
 * @example ../../docs/Message.md
 * @type { React.FC<import('types').MessageComponentProps> }
 */
const Message = (props) => {
  const {
    addNotification,
    getFlagMessageErrorNotification,
    getFlagMessageSuccessNotification,
    getMuteUserErrorNotification,
    getMuteUserSuccessNotification,
    message,
    channel: propChannel,
    onUserClick: propOnUserClick,
    onUserHover: propOnUserHover,
    openThread: propOpenThread,
    retrySendMessage: propRetrySendMessage,
    onMentionsClick: propOnMentionsClick,
    onMentionsHover: propOnMentionsHover,
    Message: MessageUIComponent = MessageSimple,
    messageActions = Object.keys(MESSAGE_ACTIONS),
    formatDate,
    groupStyles = [],
  } = props;
  const { channel: contextChannel } = useContext(ChannelContext);
  const channel = propChannel || contextChannel;
  const channelConfig = channel?.getConfig && channel.getConfig();
  const { editing, setEdit, clearEdit } = useEditHandler();
  const handleDelete = useDeleteHandler(message);
  const handleReaction = useReactionHandler(message);
  const handleAction = useActionHandler(message);
  const handleRetry = useRetryHandler(propRetrySendMessage);
  const handleOpenThread = useOpenThreadHandler(message, propOpenThread);
  const handleFlag = useFlagHandler(message, {
    notify: addNotification,
    getSuccessNotification: getFlagMessageSuccessNotification,
    getErrorNotification: getFlagMessageErrorNotification,
  });
  const handleMute = useMuteHandler(message, {
    notify: addNotification,
    getSuccessNotification: getMuteUserSuccessNotification,
    getErrorNotification: getMuteUserErrorNotification,
  });
  const { onMentionsClick, onMentionsHover } = useMentionsHandler(message, {
    onMentionsClick: propOnMentionsClick,
    onMentionsHover: propOnMentionsHover,
  });
  const { onUserClick, onUserHover } = useUserHandler(message, {
    onUserClickHandler: propOnUserClick,
    onUserHoverHandler: propOnUserHover,
  });
  const { isMyMessage, isAdmin, isModerator, isOwner } = useUserRole(message);
  const canReact = true;
  const canReply = true;
  const canEdit = isMyMessage || isModerator || isOwner || isAdmin;
  const canDelete = canEdit;
  const messageActionsHandler = useCallback(() => {
    if (!message || !messageActions) {
      return [];
    }

    return getMessageActions(messageActions, {
      canDelete,
      canEdit,
      canReply,
      canReact,
      canFlag: !isMyMessage,
      canMute: !isMyMessage && !!channelConfig?.mutes,
    });
  }, [
    channelConfig,
    message,
    messageActions,
    canDelete,
    canEdit,
    canReply,
    canReact,
    isMyMessage,
  ]);

  const actionsEnabled =
    message && message.type === 'regular' && message.status === 'received';

  return (
    MessageUIComponent && (
      <MessageUIComponent
        {...props}
        editing={editing}
        formatDate={formatDate}
        clearEditingState={clearEdit}
        setEditingState={setEdit}
        groupStyles={groupStyles}
        actionsEnabled={actionsEnabled}
        Message={MessageUIComponent}
        handleReaction={handleReaction}
        getMessageActions={messageActionsHandler}
        handleFlag={handleFlag}
        handleMute={handleMute}
        handleAction={handleAction}
        handleDelete={handleDelete}
        handleEdit={setEdit}
        handleRetry={handleRetry}
        handleOpenThread={handleOpenThread}
        isMyMessage={() => isMyMessage}
        channelConfig={channelConfig}
        onMentionsClickMessage={onMentionsClick}
        onMentionsHoverMessage={onMentionsHover}
        onUserClick={onUserClick}
        onUserHover={onUserHover}
      />
    )
  );
};

Message.propTypes = {
  /** The message object */
  message: /** @type {PropTypes.Validator<import('stream-chat').MessageResponse>} */ (PropTypes.shape(
    {
      text: PropTypes.string.isRequired,
      html: PropTypes.string.isRequired,
      type: PropTypes.string.isRequired,
      reaction_counts: PropTypes.objectOf(PropTypes.number.isRequired),
      reaction_scores: PropTypes.objectOf(PropTypes.number.isRequired),
      created_at: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
      updated_at: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
    },
  ).isRequired),
  /** The client connection object for connecting to Stream */
  client: /** @type {PropTypes.Validator<import('types').StreamChatReactClient>} */ (PropTypes.objectOf(
    checkClientPropType,
  ).isRequired),
  /** The current channel this message is displayed in */
  channel: /** @type {PropTypes.Validator<ReturnType<import('types').StreamChatReactClient['channel']>>} */ (PropTypes.objectOf(
    checkChannelPropType,
  ).isRequired),
  /** A list of users that have read this message */
  readBy: PropTypes.array,
  /** groupStyles, a list of styles to apply to this message. ie. top, bottom, single etc */
  groupStyles: PropTypes.array,

  /** Override the default formatting of the date. This is a function that has access to the original date object. Returns a string or Node  */
  formatDate: PropTypes.func,
  /**
   * Message UI component to display a message in message list.
   * Available from [channel context](https://getstream.github.io/stream-chat-react/#channelcontext)
   * */
  Message: /** @type {PropTypes.Validator<React.ElementType<import('types').MessageUIComponentProps>>} */ (PropTypes.elementType),
  /**
   * The component that will be rendered if the message has been deleted.
   * All props are passed into this component.
   */
  MessageDeleted: /** @type {PropTypes.Validator<React.ElementType<import('types').MessageDeletedProps>>} */ (PropTypes.elementType),

  /**
   * A component to display the selector that allows a user to react to a certain message.
   */
  ReactionSelector: /** @type {PropTypes.Validator<React.ElementType<import('types').ReactionSelectorProps>>} */ (PropTypes.elementType),
  /**
   * A component to display the a message list of reactions.
   */
  ReactionsList: /** @type {PropTypes.Validator<React.ElementType<import('types').ReactionsListProps>>} */ (PropTypes.elementType),
  /**
   * Attachment UI component to display attachment in individual message.
   * Available from [channel context](https://getstream.github.io/stream-chat-react/#channelcontext)
   * */
  Attachment: /** @type {PropTypes.Validator<React.ElementType<import('types').WrapperAttachmentUIComponentProps>>} */ (PropTypes.elementType),
  /** render HTML instead of markdown. Posting HTML is only allowed server-side */
  unsafeHTML: PropTypes.bool,
  /**
   * Array of allowed actions on message. e.g. ['edit', 'delete', 'mute', 'flag']
   * If all the actions need to be disabled, empty array or false should be provided as value of prop.
   * */
  messageActions: PropTypes.oneOfType([PropTypes.bool, PropTypes.array]),
  /**
   * Function that returns message/text as string to be shown as notification, when request for flagging a message is successful
   *
   * This function should accept following params:
   *
   * @param message A [message object](https://getstream.io/chat/docs/#message_format) which is flagged.
   *
   * */
  getFlagMessageSuccessNotification: PropTypes.func,
  /**
   * Function that returns message/text as string to be shown as notification, when request for flagging a message runs into error
   *
   * This function should accept following params:
   *
   * @param message A [message object](https://getstream.io/chat/docs/#message_format) which is flagged.
   *
   * */
  getFlagMessageErrorNotification: PropTypes.func,
  /**
   * Function that returns message/text as string to be shown as notification, when request for muting a user is successful
   *
   * This function should accept following params:
   *
   * @param user A user object which is being muted
   *
   * */
  getMuteUserSuccessNotification: PropTypes.func,
  /**
   * Function that returns message/text as string to be shown as notification, when request for muting a user runs into error
   *
   * This function should accept following params:
   *
   * @param user A user object which is being muted
   *
   * */
  getMuteUserErrorNotification: PropTypes.func,
  /** Latest message id on current channel */
  lastReceivedId: PropTypes.string,
  /** DOMRect object for parent MessageList component */
  messageListRect: /** @type {PropTypes.Validator<DOMRect>} */ (PropTypes.object),
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  members: /** @type {PropTypes.Validator<import('seamless-immutable').ImmutableObject<{[user_id: string]: import('stream-chat').ChannelMemberResponse<import('types').StreamChatReactUserType>}> | null | undefined>} */ (PropTypes.object),
  /**
   * Function to add custom notification on messagelist
   *
   * @param text Notification text to display
   * @param type Type of notification. 'success' | 'error'
   * */
  addNotification: PropTypes.func,
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  updateMessage: PropTypes.func,
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  removeMessage: PropTypes.func,
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  retrySendMessage: PropTypes.func,
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  onMentionsClick: PropTypes.func,
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  onMentionsHover: PropTypes.func,
  /**
   * The handler for click event on the user that posted the message
   *
   * @param event Dom click event which triggered handler.
   * @param user the User object for the corresponding user.
   */
  onUserClick: PropTypes.func,
  /**
   * The handler for hover events on the user that posted the message
   *
   * @param event Dom hover event which triggered handler.
   * @param user the User object for the corresponding user.
   */
  onUserHover: PropTypes.func,
  /** @see See [Channel Context](https://getstream.github.io/stream-chat-react/#channelcontext) */
  openThread: PropTypes.func,
  /**
   * Additional props for underlying MessageInput component.
   * Available props - https://getstream.github.io/stream-chat-react/#messageinput
   * */
  additionalMessageInputProps: PropTypes.object,
};

Message.defaultProps = {
  Message: MessageSimple,
  readBy: [],
  groupStyles: [],
  messageActions: Object.keys(MESSAGE_ACTIONS),
};

export default React.memo(Message, areMessagePropsEqual);
