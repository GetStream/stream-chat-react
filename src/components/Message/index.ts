export * from './FixedHeightMessage';
export * from './hooks';
export * from './icons';
export * from './Message';
export * from './MessageDeleted';
export * from './MessageOptions';
export * from './MessageRepliesCountButton';
export * from './MessageSimple';
export * from './MessageStatus';
export * from './MessageText';
export * from './MessageTimestamp';
export * from './QuotedMessage';
export * from './types';
export * from './utils';
