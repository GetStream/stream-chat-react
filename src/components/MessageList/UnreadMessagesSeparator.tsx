import React from 'react';
import { useTranslationContext } from '../../context';

export const UNREAD_MESSAGE_SEPARATOR_CLASS = 'str-chat__unread-messages-separator';

export type UnreadMessagesSeparatorProps = {
  /**
   * Configuration parameter to determine, whether the unread count is to be shown on the component. Disabled by default.
   */
  showCount?: boolean;
  /**
   * The count of unread messages to be displayed if enabled.
   */
  unreadCount?: number;
};

export const UnreadMessagesSeparator = ({
  showCount,
  unreadCount,
}: UnreadMessagesSeparatorProps) => {
  const { t } = useTranslationContext('UnreadMessagesSeparator');
  return (
    <div className={UNREAD_MESSAGE_SEPARATOR_CLASS} data-testid='unread-messages-separator'>
      {unreadCount && showCount
        ? t<string>('unreadMessagesSeparatorText', { count: unreadCount })
        : t<string>('Unread messages')}
    </div>
  );
};
