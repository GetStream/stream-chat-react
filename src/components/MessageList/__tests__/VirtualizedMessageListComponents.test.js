import React from 'react';
import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import {
  EmptyPlaceholder,
  Footer,
  Header,
  Item,
  messageRenderer,
} from '../VirtualizedMessageListComponents';
import {
  generateChannel,
  generateMessage,
  generateUser,
  getTestClientWithUser,
  initClientWithChannels,
} from '../../../mock-builders';
import {
  ChannelActionProvider,
  ChannelStateProvider,
  ChatProvider,
  ComponentProvider,
  TranslationProvider,
} from '../../../context';
import { MessageSimple } from '../../Message';
import { UnreadMessagesSeparator } from '../UnreadMessagesSeparator';

const prependOffset = 0;
const user1 = generateUser();
const user2 = generateUser();
let client;
let channel;

const PREPEND_OFFSET = 10 ** 7;

const Wrapper = ({ children, componentContext = {} }) => (
  <ChatProvider value={{ client }}>
    <ChannelStateProvider value={{ channel }}>
      <ChannelActionProvider value={{ addNotification: jest.fn() }}>
        <ComponentProvider value={componentContext}>{children}</ComponentProvider>
      </ChannelActionProvider>
    </ChannelStateProvider>
  </ChatProvider>
);

const renderElements = (children, componentContext) =>
  render(<Wrapper componentContext={componentContext}>{children}</Wrapper>);

describe('VirtualizedMessageComponents', () => {
  describe('Item', function () {
    const processedMessages = [generateMessage()];
    const withVirtualMessageClasses = { virtualMessage: 'XXX' };
    const withMessageGroupStyles = { [processedMessages[0].id]: 'single' };
    const withoutVirtualMessageClasses = undefined;
    const withoutMessageGroupStyles = {};

    it.each([
      ['with', 'without', withVirtualMessageClasses, withoutMessageGroupStyles],
      ['without', 'without', withoutVirtualMessageClasses, withoutMessageGroupStyles],
      ['without', 'with', withoutVirtualMessageClasses, withMessageGroupStyles],
      ['with', 'with', withVirtualMessageClasses, withMessageGroupStyles],
    ])(
      'should render wrapper %s custom classes %s group styles',
      (_, __, customClasses, messageGroupStyles) => {
        const props = {
          'data-item-index': PREPEND_OFFSET,
        };
        const virtuosoContext = {
          customClasses,
          messageGroupStyles,
          numItemsPrepended: 0,
          processedMessages,
        };

        const { container } = renderElements(<Item context={virtuosoContext} {...props} />);
        expect(container).toMatchSnapshot();
      },
    );
  });

  describe('Header', () => {
    const head = <div>Custom head</div>;
    const CustomLoadingIndicator = () => <div>Custom Loading Indicator</div>;
    it('should render empty div in Header when not loading more messages', () => {
      const { container } = renderElements(<Header />);
      expect(container).toMatchInlineSnapshot(`<div />`);
    });

    it('should render LoadingIndicator in Header when loading more messages', () => {
      const context = { loadingMore: true };
      const { container } = renderElements(<Header context={context} />);
      expect(container).toMatchSnapshot();
    });

    it('should render custom LoadingIndicator in Header when loading more messages', () => {
      const componentContext = { LoadingIndicator: CustomLoadingIndicator };
      const context = { loadingMore: true };
      const { container } = renderElements(<Header context={context} />, componentContext);
      expect(container).toMatchInlineSnapshot(`
        <div>
          <div
            class="str-chat__virtual-list__loading"
          >
            <div>
              Custom Loading Indicator
            </div>
          </div>
        </div>
      `);
    });

    it('should not render custom LoadingIndicator in Header when not loading more messages', () => {
      const componentContext = { LoadingIndicator: CustomLoadingIndicator };
      const { container } = renderElements(<Header />, componentContext);
      expect(container).toMatchInlineSnapshot(`<div />`);
    });

    // FIXME: this is a crazy pattern of having to set LoadingIndicator to null so that additionalVirtuosoProps.head can be rendered.
    it('should not render custom head in Header when loading more messages, but the LoadingIndicator', () => {
      const context = { head, loadingMore: true };
      const { container } = renderElements(<Header context={context} />);
      expect(container).toMatchSnapshot();
    });

    // FIXME: this is a crazy pattern of having to set LoadingIndicator to null so that additionalVirtuosoProps.head can be rendered.
    it('should render custom head in Header when LoadingIndicator in component context is set to null', () => {
      const componentContext = {
        LoadingIndicator: null,
      };
      const context = { head, loadingMore: true };
      const { container } = renderElements(<Header context={context} />, componentContext);
      expect(container).toMatchInlineSnapshot(`
        <div>
          <div>
            Custom head
          </div>
        </div>
      `);
    });

    it('should not render custom head in Header when not loading more messages', () => {
      const context = { head };
      const { container } = renderElements(<Header context={context} />);
      expect(container).toMatchInlineSnapshot(`<div />`);
    });

    it('should render custom LoadingIndicator instead of head when loading more', () => {
      const componentContext = { LoadingIndicator: CustomLoadingIndicator };
      const context = { head, loadingMore: true };
      const { container } = renderElements(<Header context={context} />, componentContext);
      expect(container).toMatchInlineSnapshot(`
        <div>
          <div
            class="str-chat__virtual-list__loading"
          >
            <div>
              Custom Loading Indicator
            </div>
          </div>
        </div>
      `);
    });
  });

  describe('EmptyPlaceholder', () => {
    const EmptyStateIndicator = ({ listType }) => (
      <div data-listType={listType}>Custom EmptyStateIndicator</div>
    );
    const NullEmptyStateIndicator = null;
    const componentContext = { EmptyStateIndicator };
    it('should render for main message list by default', () => {
      const { container } = renderElements(<EmptyPlaceholder />);
      expect(container).toMatchSnapshot();
    });

    it('should render empty for thread by default', () => {
      const { container } = renderElements(<EmptyPlaceholder context={{ threadList: true }} />);
      expect(container).toMatchInlineSnapshot(`<div />`);
    });
    it('should render custom EmptyStateIndicator for main message list', () => {
      const { container } = renderElements(<EmptyPlaceholder />, componentContext);
      expect(container).toMatchSnapshot();
    });

    it('should render custom EmptyStateIndicator for thread', () => {
      const { container } = renderElements(
        <EmptyPlaceholder context={{ threadList: true }} />,
        componentContext,
      );
      expect(container).toMatchSnapshot();
    });

    it('should render empty if EmptyStateIndicator nullified', () => {
      const componentContext = { EmptyStateIndicator: NullEmptyStateIndicator };
      const { container } = renderElements(<EmptyPlaceholder />, componentContext);
      expect(container).toMatchInlineSnapshot(`<div />`);
    });

    it('should render empty in thread if EmptyStateIndicator nullified', () => {
      const componentContext = { EmptyStateIndicator: NullEmptyStateIndicator };
      const { container } = renderElements(
        <EmptyPlaceholder context={{ threadList: true }} />,
        componentContext,
      );
      expect(container).toMatchInlineSnapshot(`<div />`);
    });
  });

  describe('Footer', () => {
    it('should render nothing in Footer by default', () => {
      const { container } = renderElements(<Footer />);
      expect(container).toMatchInlineSnapshot(`<div />`);
    });
    it('should render custom TypingIndicator in Footer', () => {
      const TypingIndicator = () => <div>Custom TypingIndicator</div>;
      const componentContext = { TypingIndicator };
      const { container } = renderElements(<Footer />, componentContext);
      expect(container).toMatchInlineSnapshot(`
        <div>
          <div>
            Custom TypingIndicator
          </div>
        </div>
      `);
    });
  });

  describe('messageRenderer', () => {
    const virtuosoIndex = PREPEND_OFFSET;
    const numItemsPrepended = 0;
    beforeAll(async () => {
      client = await getTestClientWithUser();
      const channelData = generateChannel();
      channel = client.channel(channelData.channel.type, channelData.channel.id, channelData);
    });

    it('should allow to execute custom item rendering logic instead of the default', () => {
      const customMessageRenderer = jest.fn();
      const virtuosoContext = {
        customMessageRenderer,
        numItemsPrepended,
        processedMessages: [generateMessage()],
      };
      messageRenderer(virtuosoIndex, undefined, virtuosoContext);
      expect(customMessageRenderer).toHaveBeenCalledWith(
        expect.arrayContaining(virtuosoContext.processedMessages),
        0,
      );
    });

    describe('default item rendering logic', () => {
      it('should render nothing if MessageSystem component is undefined for system messages', () => {
        const virtuosoContext = {
          numItemsPrepended,
          processedMessages: [generateMessage({ type: 'system' })],
        };

        const { container } = render(messageRenderer(virtuosoIndex, undefined, virtuosoContext));
        expect(container).toMatchInlineSnapshot(`<div />`);
      });

      it('should render MessageSystem component for system messages', () => {
        const text = 'MessageSystem';
        const MessageSystem = () => <>{text}</>;
        const virtuosoContext = {
          MessageSystem,
          numItemsPrepended,
          processedMessages: [generateMessage({ type: 'system' })],
        };

        render(messageRenderer(virtuosoIndex, undefined, virtuosoContext));
        expect(screen.getByText(text)).toBeInTheDocument();
      });

      it('should render nothing if DateSeparator component is undefined for custom message type date', () => {
        const virtuosoContext = {
          numItemsPrepended,
          processedMessages: [generateMessage({ customType: 'message.date', date: new Date() })],
        };
        const { container } = render(messageRenderer(virtuosoIndex, undefined, virtuosoContext));
        expect(container).toMatchInlineSnapshot(`<div />`);
      });

      it('should render DateSeparator component for custom message type date', () => {
        const text = 'DateSeparator';
        const DateSeparator = () => <div>{text}</div>;
        const virtuosoContext = {
          DateSeparator,
          numItemsPrepended,
          processedMessages: [generateMessage({ customType: 'message.date', date: new Date() })],
        };
        render(messageRenderer(virtuosoIndex, undefined, virtuosoContext));
        expect(screen.getByText(text)).toBeInTheDocument();
      });

      it('should render empty div when trying to render message at non-existent index', () => {
        const virtuosoIndex = PREPEND_OFFSET - 1;
        const virtuosoContext = {
          numItemsPrepended,
          processedMessages: [generateMessage()],
        };

        const { container } = render(messageRenderer(virtuosoIndex, undefined, virtuosoContext));
        expect(container).toMatchInlineSnapshot(`
                  <div>
                    <div
                      style="height: 1px;"
                    />
                  </div>
              `);
      });

      describe('UnreadMessagesSeparator', () => {
        const messages = Array.from({ length: 2 }, (_, i) =>
          generateMessage({ created_at: new Date(i).toISOString(), id: i + 1 }),
        );

        const Message = () => <div className='message-component' />;

        const renderMarkUnread = async ({ virtuosoContext, virtuosoIndex } = {}) => {
          const {
            channels: [channel],
            client,
          } = await initClientWithChannels();
          return render(
            <ChatProvider value={{ client }}>
              <TranslationProvider value={{ t: (v) => v }}>
                <ComponentProvider value={{}}>
                  <ChannelActionProvider value={{}}>
                    <ChannelStateProvider value={{ channel }}>
                      {messageRenderer(virtuosoIndex ?? PREPEND_OFFSET, undefined, virtuosoContext)}
                    </ChannelStateProvider>
                  </ChannelActionProvider>
                </ComponentProvider>
              </TranslationProvider>
            </ChatProvider>,
          );
        };

        it('should be rendered below the last read message if unread count is non-zero', async () => {
          const { container } = await renderMarkUnread({
            virtuosoContext: {
              lastReadMessageId: messages[0].id,
              lastReceivedMessageId: messages[1].id,
              Message,
              numItemsPrepended,
              ownMessagesReadByOthers: {},
              processedMessages: messages,
              unreadMessageCount: 1,
              UnreadMessagesSeparator,
              virtuosoRef: { current: {} },
            },
          });
          expect(container).toMatchInlineSnapshot(`
            <div>
              <div
                class="message-component"
              />
              <div
                class="str-chat__unread-messages-separator-wrapper"
              >
                <div
                  class="str-chat__unread-messages-separator"
                  data-testid="unread-messages-separator"
                >
                  Unread messages
                </div>
              </div>
            </div>
          `);
        });

        it('should not be rendered below the last read message if the message is the newest in the channel', async () => {
          const { container } = await renderMarkUnread({
            virtuosoContext: {
              lastReadMessageId: messages[1].id,
              lastReceivedMessageId: messages[1].id,
              Message,
              numItemsPrepended: 1,
              ownMessagesReadByOthers: {},
              processedMessages: messages,
              unreadMessageCount: 1,
              UnreadMessagesSeparator,
              virtuosoRef: { current: {} },
            },
          });
          expect(container).toMatchInlineSnapshot(`
            <div>
              <div
                class="message-component"
              />
            </div>
          `);
        });

        it('should be rendered if unread count is falsy and first unread message is known', async () => {
          const { container } = await renderMarkUnread({
            virtuosoContext: {
              firstUnreadMessageId: messages[1].id,
              lastReadMessageId: messages[0].id,
              lastReceivedMessageId: messages[1].id,
              Message,
              numItemsPrepended,
              ownMessagesReadByOthers: {},
              processedMessages: messages,
              unreadMessageCount: 0,
              UnreadMessagesSeparator,
              virtuosoRef: { current: {} },
            },
          });
          expect(container).toMatchInlineSnapshot(`
            <div>
              <div
                class="message-component"
              />
              <div
                class="str-chat__unread-messages-separator-wrapper"
              >
                <div
                  class="str-chat__unread-messages-separator"
                  data-testid="unread-messages-separator"
                >
                  Unread messages
                </div>
              </div>
            </div>
          `);
        });

        it('should not be rendered if unread count is falsy and first unread messages is unknown', async () => {
          const { container } = await renderMarkUnread({
            virtuosoContext: {
              lastReadMessageId: messages[0].id,
              lastReceivedMessageId: messages[1].id,
              Message,
              numItemsPrepended,
              ownMessagesReadByOthers: {},
              processedMessages: messages,
              unreadMessageCount: 0,
              UnreadMessagesSeparator,
              virtuosoRef: { current: {} },
            },
          });
          expect(container).toMatchInlineSnapshot(`
            <div>
              <div
                class="message-component"
              />
            </div>
          `);
        });

        it('should not be rendered if rendering other message than the last read one', async () => {
          const { container } = await renderMarkUnread({
            virtuosoContext: {
              lastReadMessageId: messages[0].id,
              lastReceivedMessageId: messages[1].id,
              Message,
              numItemsPrepended: 1,
              ownMessagesReadByOthers: {},
              processedMessages: messages,
              unreadMessageCount: 1,
              UnreadMessagesSeparator,
              virtuosoRef: { current: {} },
            },
          });
          expect(container).toMatchInlineSnapshot(`
            <div>
              <div
                class="message-component"
              />
            </div>
          `);
        });
      });

      it.each([
        ['not ', 'by default', 'not ', false],
        ['', '', '', true],
      ])(
        'should %sgroup messages %s and mark the first and the last group message',
        (_, __, ___, shouldGroupByUser) => {
          const virtuosoRef = { current: {} };
          const user1MessageGroup = [
            generateMessage({ user: user1 }),
            generateMessage({ user: user1 }),
            generateMessage({ user: user1 }),
            generateMessage({ user: user1 }),
          ];
          const processedMessages = [
            generateMessage({ user: user2 }),
            ...user1MessageGroup,
            generateMessage({ user: user2 }),
          ];

          const { container } = renderElements(
            <>
              {processedMessages.map((_, numItemsPrepended) => {
                const virtuosoContext = {
                  Message: MessageSimple,
                  numItemsPrepended,
                  ownMessagesReadByOthers: {},
                  prependOffset,
                  processedMessages,
                  shouldGroupByUser,
                  virtuosoRef,
                };
                return (
                  <div key={numItemsPrepended}>
                    {messageRenderer(virtuosoIndex, undefined, virtuosoContext)}
                  </div>
                );
              })}
            </>,
          );
          const messageElements = container.getElementsByClassName(
            'str-chat__message str-chat__message-simple',
          );

          const firstGroupItemClass = 'str-chat__virtual-message__wrapper--first';
          const lastGroupItemClass = 'str-chat__virtual-message__wrapper--end';
          expect(
            container.getElementsByClassName('str-chat__virtual-message__wrapper--group'),
          ).toHaveLength(shouldGroupByUser ? user1MessageGroup.length - 1 : 0);

          expect(container.getElementsByClassName(firstGroupItemClass)).toHaveLength(
            shouldGroupByUser ? 3 : 0,
          );
          expect(container.getElementsByClassName(lastGroupItemClass)).toHaveLength(
            shouldGroupByUser ? 3 : 0,
          );
          if (shouldGroupByUser) {
            expect(messageElements[0]).toHaveClass(firstGroupItemClass);
            expect(messageElements[0]).toHaveClass(lastGroupItemClass);
            expect(messageElements[1]).toHaveClass(firstGroupItemClass);
            expect(messageElements[processedMessages.length - 2]).toHaveClass(lastGroupItemClass);
            expect(messageElements[processedMessages.length - 1]).toHaveClass(firstGroupItemClass);
            expect(messageElements[processedMessages.length - 1]).toHaveClass(lastGroupItemClass);
          }
        },
      );
    });
  });
});
