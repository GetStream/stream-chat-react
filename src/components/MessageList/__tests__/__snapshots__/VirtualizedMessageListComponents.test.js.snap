// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VirtualizedMessageComponents EmptyPlaceholder should render custom EmptyStateIndicator for main message list 1`] = `
<div>
  <div
    data-listtype="message"
  >
    Custom EmptyStateIndicator
  </div>
</div>
`;

exports[`VirtualizedMessageComponents EmptyPlaceholder should render custom EmptyStateIndicator for thread 1`] = `
<div>
  <div
    data-listtype="thread"
  >
    Custom EmptyStateIndicator
  </div>
</div>
`;

exports[`VirtualizedMessageComponents EmptyPlaceholder should render for main message list by default 1`] = `
<div>
  <div
    class="str-chat__empty-channel"
  >
    <svg
      data-testid="chat-bubble"
      fill="none"
      height="96"
      viewBox="0 0 136 136"
      width="96"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M106 24.5H30C24.775 24.5 20.5 28.775 20.5 34V119.5L39.5 100.5H106C111.225 100.5 115.5 96.225 115.5 91V34C115.5 28.775 111.225 24.5 106 24.5ZM106 91H39.5L30 100.5V34H106V91Z"
        fill="#B4B7BB"
      />
    </svg>
    <p
      class="str-chat__empty-channel-text"
      role="listitem"
    >
      No chats here yetâ€¦
    </p>
  </div>
</div>
`;

exports[`VirtualizedMessageComponents Header should not render custom head in Header when loading more messages, but the LoadingIndicator 1`] = `
<div>
  <div
    class="str-chat__virtual-list__loading"
  >
    <div
      class="str-chat__loading-indicator"
      data-testid="loading-indicator-wrapper"
      style="height: 20px; width: 20px;"
    >
      <svg
        height="20"
        viewBox="0 0 30 30"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <lineargradient
            id="a"
            x1="50%"
            x2="50%"
            y1="0%"
            y2="100%"
          >
            <stop
              offset="0%"
              stop-color="#FFF"
              stop-opacity="0"
            />
            <stop
              data-testid="loading-indicator-circle"
              offset="100%"
              stop-color="#006CFF"
              stop-opacity="1"
              style="stop-color: rgb(0, 108, 255);"
            />
          </lineargradient>
        </defs>
        <path
          d="M2.518 23.321l1.664-1.11A12.988 12.988 0 0 0 15 28c7.18 0 13-5.82 13-13S22.18 2 15 2V0c8.284 0 15 6.716 15 15 0 8.284-6.716 15-15 15-5.206 0-9.792-2.652-12.482-6.679z"
          fill="url(#a)"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`VirtualizedMessageComponents Header should render LoadingIndicator in Header when loading more messages 1`] = `
<div>
  <div
    class="str-chat__virtual-list__loading"
  >
    <div
      class="str-chat__loading-indicator"
      data-testid="loading-indicator-wrapper"
      style="height: 20px; width: 20px;"
    >
      <svg
        height="20"
        viewBox="0 0 30 30"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <defs>
          <lineargradient
            id="a"
            x1="50%"
            x2="50%"
            y1="0%"
            y2="100%"
          >
            <stop
              offset="0%"
              stop-color="#FFF"
              stop-opacity="0"
            />
            <stop
              data-testid="loading-indicator-circle"
              offset="100%"
              stop-color="#006CFF"
              stop-opacity="1"
              style="stop-color: rgb(0, 108, 255);"
            />
          </lineargradient>
        </defs>
        <path
          d="M2.518 23.321l1.664-1.11A12.988 12.988 0 0 0 15 28c7.18 0 13-5.82 13-13S22.18 2 15 2V0c8.284 0 15 6.716 15 15 0 8.284-6.716 15-15 15-5.206 0-9.792-2.652-12.482-6.679z"
          fill="url(#a)"
          fill-rule="evenodd"
        />
      </svg>
    </div>
  </div>
</div>
`;

exports[`VirtualizedMessageComponents Item should render wrapper with custom classes with group styles 1`] = `
<div>
  <div
    class="XXX"
    data-item-index="10000000"
  />
</div>
`;

exports[`VirtualizedMessageComponents Item should render wrapper with custom classes without group styles 1`] = `
<div>
  <div
    class="XXX"
    data-item-index="10000000"
  />
</div>
`;

exports[`VirtualizedMessageComponents Item should render wrapper without custom classes with group styles 1`] = `
<div>
  <div
    class="str-chat__virtual-list-message-wrapper str-chat__li str-chat__li--single"
    data-item-index="10000000"
  />
</div>
`;

exports[`VirtualizedMessageComponents Item should render wrapper without custom classes without group styles 1`] = `
<div>
  <div
    class="str-chat__virtual-list-message-wrapper str-chat__li"
    data-item-index="10000000"
  />
</div>
`;
